students/yixingxu/lesson05/mailroom3.py

```diff
@@ -0,0 +1,145 @@
+        f"Thanks for your generous donation of ${donation_amount}"
+        )
+
+# construct list of names
+def list_name(donation_history = donation_history):
+    namelist = []
+    for item in donation_history:
+        namelist.append(item['name'])
+    return namelist     
+
+# add name to the list    
+def add_name(name,donation_history = donation_history):
+    donation_history.append({'name':name, 'donations':[]})
+    print('\nNew name {} has been added to the list'.format(name))
+    
+def add_donation_amount(name, donation_amount, donation_history = donation_history):
+    for item in donation_history:
+        if item['name'] == name:
+            item['donations'].append(donation_amount)
+            break
+    
+def create_a_report(donation_history=donation_history):
+    # create a summary list
+    summary_list = create_summary()
+    # sort summary list according to the total donation
+    sorted_list = summary_list[:]
+    sort_list(sorted_list)
+    print_report(sorted_list)
+    
+# print a list
+def print_report(list):
+    Donor_name_width = 20
+    Total_given_width = 15
+    Num_gifts_width = 5
+    Average_gift_width = 15
+    seperator_width = 2
+    title = ['Donor Name','|','Total Given', '|', 'Num Gifts', '|', 'Average Gift']
+    divider = '-'*(Donor_name_width+Total_given_width+Num_gifts_width+Average_gift_width+seperator_width*5)
+    print('{:<20}{:<2}{:<15}{:<2}{:<5}{:<2}{:<15}'.format(*title))
+    print(divider)
+    for item in list:
+        print('{:<20}{:2}{:15.2f}  {:8} {:>2}{:12.2f}\n'.format(*item))
+     
+# create a summary list
+def create_summary(donation_history=donation_history):
+    summary_list = []
+    for item in donation_history:
+        total = sum(item['donations'])
+        number_of_gifts = len(item['donations'])
+        average = total/number_of_gifts
+        summary_list.append([item['name'],'$',total,number_of_gifts,'$',average])
+    return summary_list
+
+# sort summary list according to the total donation
+def sort_list(sorted_list):
+    #return summary_list.sort(key=sort_by_total_given)
+    sorted_list.sort(key=sort_by_total_given,reverse=True)
+
+# get the total given amount 
+def sort_by_total_given(list):
+    return list[2]
+    
+# quit
+def Quit():
+    print('Quit')
+    return 'exit menu'
+
+def send_letters_to_everyone(donation_history = donation_history):
+    for item in donation_history:     
+        Letter = "Dear {name},\n\n        Thank you for all your very kind donations of ${donations}\n\n        It will be put to very good use.\n\n                       Sincerely,\n                          -The Team".format(**item)
+        with open(item['name']+'.txt', 'w') as wf:
+            wf.write(Letter)
+    print("\nAll letters written to files")
+   
+    
+if __name__ == "__main__":
+    main_dispatch = {
+                '1': send_a_thank_you,
+                '2': create_a_report,
+                '3': send_letters_to_everyone,
+                '4': Quit
```

Comment:
should be lower case variable and to avoid colliding with built-in `quit` you can name it `quit_program`